internal/controller/metalstackfirewalldeployment_controller.go

```diff
@@ -0,0 +1,471 @@
+
+	var (
+		name = fmt.Sprintf("%s-firewall", r.infraCluster.GetName())
+		tags = []string{
+			tag.New(tag.ClusterID, r.infraCluster.Spec.NodeNetworkID),
+			tag.New(v1alpha1.TagInfraClusterResource, fmt.Sprintf("%s.%s", r.infraCluster.Namespace, r.infraCluster.Name)),
+			tag.New(fcmv2.FirewallControllerManagedByAnnotation, "cluster-api-provider-metal-stack"),
+			tag.New((v1alpha1.TagFirewallDeploymentResource), fmt.Sprintf("%s.%s", r.firewallDeployment.Namespace, r.firewallDeployment.Name)),
+		}
+	)
+
+	if r.firewallDeployment.Spec.ManagedResourceRef != nil {
+		return nil
+	}
+
+	networkIDs := r.firewallTemplate.Spec.Networks
+	if !slices.Contains(networkIDs, r.infraCluster.Spec.NodeNetworkID) {
+		networkIDs = append(networkIDs, r.infraCluster.Spec.NodeNetworkID)
+	}
+
+	networks := make([]*models.V1MachineAllocationNetwork, 0, len(networkIDs))
+	for _, n := range networkIDs {
+		network := &models.V1MachineAllocationNetwork{
+			Networkid:   &n,
+			Autoacquire: ptr.To(true),
+		}
+		networks = append(networks, network)
+	}
+
+	if r.firewallTemplate.Spec.InitialRuleSet == nil {
+		return fmt.Errorf("firewall template %s/%s has no initial rule set defined and will not allow any traffic", r.firewallTemplate.Namespace, r.firewallTemplate.Name)
+	}
+
+	egressRules := make([]*models.V1FirewallEgressRule, 0, len(r.firewallTemplate.Spec.InitialRuleSet.Egress))
+	for _, er := range r.firewallTemplate.Spec.InitialRuleSet.Egress {
+		egressRules = append(egressRules, &models.V1FirewallEgressRule{
+			Comment:  er.Comment,
+			Ports:    er.Ports,
+			Protocol: string(er.Protocol),
+			To:       er.To,
+		})
+	}
+
+	ingressRules := make([]*models.V1FirewallIngressRule, 0, len(r.firewallTemplate.Spec.InitialRuleSet.Ingress))
+	for _, ir := range r.firewallTemplate.Spec.InitialRuleSet.Ingress {
+		ingressRules = append(ingressRules, &models.V1FirewallIngressRule{
+			Comment:  ir.Comment,
+			Ports:    ir.Ports,
+			Protocol: string(ir.Protocol),
+			From:     ir.From,
+		})
+	}
+
+	fwresp, err := r.metalClient.Firewall().AllocateFirewall(firewall.NewAllocateFirewallParamsWithContext(r.ctx).WithBody(&models.V1FirewallCreateRequest{
+		Hostname:    name,
+		Name:        name,
+		Description: fmt.Sprintf("firewall for cluster %s", r.cluster.GetName()),
+		Partitionid: ptr.To(r.infraCluster.Spec.Partition),
+		Projectid:   &r.infraCluster.Spec.ProjectID,
+		Tags:        tags,
+		SSHPubKeys:  []string{},
+		Networks:    networks,
+		Imageid:     ptr.To(r.firewallTemplate.Spec.Image),
+		Sizeid:      ptr.To(r.firewallTemplate.Spec.Size),
+		FirewallRules: &models.V1FirewallRules{
+			Egress:  egressRules,
+			Ingress: ingressRules,
+		},
+	}), nil)
+	if err != nil {
+		return fmt.Errorf("error creating firewall deployment: %w", err)
+	}
+
+	r.log.Info("created firewall deployment", "firewallID", fwresp.Payload.ID)
+
+	r.firewallDeployment.Spec.ManagedResourceRef = &v1alpha1.MetalStackManagedResourceRef{
```

Comment:
This approach could lead to leaking resources when storing the managed resource ref goes wrong (because there is no lookup if the firewall was already created). I guess it's just temporary anyway so we can address it later when we replace it with the firewall deployment?

Reply:
Jup, this is just temporary. Later we will lookup the firewall deployment resource.